### PR TITLE
Add node-exclusive scheduling

### PR DESCRIPTION
**Pull request by dongahn**
This PR adds node-exclusive scheduling support into fluxion.
It includes core algorithms for node-exclusive scheduling and
augment various existing tests to make sure the new support
works as intended for various existing functionality.

The main idea of node-exclusive scheduling enablement is
through shadowed resource emission support.

The core problem with adding node-exclusive scheduling is
that we did not have an ability to emit all of the node-local
resources (e.g., core, gpu, etc) when a node is exclusively
allocated.  We used to emit only those resources that are
explicitly matched with the resource types as specified
in the jobspec while leaving those unmatched resources
alone even those they are under an exclusively allocated
resource (e.g., node). We refer those unmatched resources
as "shadowed resources." 

This PR solves that problem by adding some algorithmic
changes to emit those shadowed resources.

To make this feature extensible, this allows users to
register one or more "exclusive" resource types beyond
the "node" type.

This PR also adds two node exclusive match policies:
1) low ID node first with node exclusivity (`lonodex`);
and 2) high ID node first with node exclusivity (`hinodex`).
They register the "node" type as the only exclusive
resource type and extend `lonode` and `hinode` policies
respectively.

Add a few bug fixes as well.

### CODE REVIEW

**Comment by grondo** [comment]
@dongahn: I gave this a sanity check within a job on fluke by building flux-sched and then executing a `flux mini alloc` like so:

```console
grondo@fluke108:~/git/flux-sched.git $ flux mini alloc \
    --env=FLUX_MODULE_PATH_PREPEND=$(pwd)/qmanager/modules/.libs:$(pwd)/resource/modules/.libs \
    --env=FLUXION_RESOURCE_OPTIONS="policy=lonodex" -n 16 
```

Basic, single task jobs seem to work as expected
```
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git $ flux mini submit sleep 120
ƒ666z3U3
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git $ flux job info ƒ666z3U3 R
{"version": 1, "execution": {"R_lite": [{"rank": "0", "children": {"core": "0-3"}}], "nodelist": ["fluke100"], "starttime": 1642788658, "expiration": 1643393458}}

ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux mini submit -c2 sleep 120
ƒ2npvKnib
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux job info ƒ2npvKnib R
{"version": 1, "execution": {"R_lite": [{"rank": "0", "children": {"core": "0-3"}}], "nodelist": ["fluke100"], "starttime": 1642788882, "expiration": 1643393682}}
```


But when I request 2 tasks, the assigned resource set seems to unexpectedly have 2 nodes:
```
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux mini submit -n2 sleep 120 ƒ31YnxLej
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux job info ƒ31YnxLej R
{"version": 1, "execution": {"R_lite": [{"rank": "1-2", "children": {"core": "0-3"}}], "nodelist": ["fluke[101-102]"], "starttime": 1642788911, "expiration": 1643393711}}
```

Maybe this is intended behavior? With node-exclusive scheduling does each task slot get assigned a node exclusively? In a way that does make sense for batch scripts, where a task slot is always for one broker. However would this prevent users from making requests like "I need enough resources for 4 task slots with 2 cores each" with node-exclusive resource assignment? The expectation might be that I would get 2 nodes on fluke (where each node has 4 cores), but in the current design I think we'd get 4 nodes...

**Comment by dongahn** [comment]
Thanks @grondo.

>But when I request 2 tasks, the assigned resource set seems to unexpectedly have 2 nodes:

Hmmm. This should be scheduled to a single node. I will take a look.

**Comment by dongahn** [comment]
@grondo: I will get to this myself this afternoon but would the following makes any difference?

```
--env=FLUXION_RESOURCE_OPTIONS="policy=lonodex load-allowlist=node,core,gpu"
```

**Comment by grondo** [comment]
@dongahn, I will try that when I get a chance. My house just lost power and
we're not expected to be back online until after 3pm.

**Comment by grondo** [comment]
>  I will get to this myself this afternoon but would the following makes any difference?
> `--env=FLUXION_RESOURCE_OPTIONS="policy=lonodex load-allowlist=node,core,gpu"`

@dongahn - I _think_ I tried that to no avail:

```
 grondo@fluke108:~/git/flux-sched.git$ flux mini alloc --env=FLUX_MODULE_PATH_PREPEND=$(pwd)/qmanager/modules/.libs:$(pwd)/resource/modules/.libs --env=FLUXION_RESOURCE_OPTIONS="policy=lonodex load-allowlist=node,core,gpu" -n 16
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux resource list
     STATE NNODES   NCORES    NGPUS NODELIST
      free      4       16        0 fluke[100-103]
 allocated      0        0        0 
      down      0        0        0 
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux mini run -n2 hostname
fluke100
fluke101
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux mini submit --cc=1-2 --watch hostname
ƒFGvdvjy
ƒFGvdvjz
fluke100
fluke101
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux jobs -a
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME NODELIST
    ƒFGvdvjz grondo   hostname   CD      1      1   0.049s fluke101
    ƒFGvdvjy grondo   hostname   CD      1      1   0.045s fluke100
    ƒ5mSj5kK grondo   hostname   CD      2      2   0.052s fluke[100-101]
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux job info ƒ5mSj5kK R
{"version": 1, "execution": {"R_lite": [{"rank": "0-1", "children": {"core": "0-3"}}], "nodelist": ["fluke[100-101]"], "starttime": 1642800633, "expiration": 1643405433}}
```

Is there some way to verify that the `load-allowlist` was set as requested?

Here is `flux dmesg` output in case it is useful:
```
ƒ(s=4,d=1) grondo@fluke100:~/git/flux-sched.git$ flux dmesg | grep fluxion
2022-01-21T21:30:22.535553Z broker.info[0]: rc1.0: running /etc/flux/rc1.d/01-sched-fluxion-resource-start
2022-01-21T21:30:22.683233Z broker.debug[0]: insmod sched-fluxion-resource
2022-01-21T21:30:22.683640Z sched-fluxion-resource.debug[0]: mod_main: resource module starting
2022-01-21T21:30:22.691973Z broker.info[0]: rc1.0: running /etc/flux/rc1.d/02-sched-fluxion-qmanager-start
2022-01-21T21:30:22.738116Z broker.debug[0]: insmod sched-fluxion-qmanager
2022-01-21T21:30:22.738588Z sched-fluxion-qmanager.debug[0]: service_register
2022-01-21T21:30:22.738611Z sched-fluxion-qmanager.debug[0]: enforced policy (queue=default): fcfs
2022-01-21T21:30:22.738616Z sched-fluxion-qmanager.debug[0]: effective queue params (queue=default): default
2022-01-21T21:30:22.738618Z sched-fluxion-qmanager.debug[0]: effective policy params (queue=default): default
2022-01-21T21:30:23.748924Z sched-fluxion-resource.info[0]: populate_resource_db: loaded resources from core's resource.acquire
2022-01-21T21:30:23.749082Z sched-fluxion-resource.debug[0]: resource status changed (rankset=[all] status=DOWN)
2022-01-21T21:30:23.749093Z sched-fluxion-resource.debug[0]: mod_main: resource graph database loaded
2022-01-21T21:30:23.749433Z sched-fluxion-qmanager.debug[0]: handshaking with sched-fluxion-resource completed
2022-01-21T21:30:23.749135Z sched-fluxion-resource.debug[0]: resource status changed (rankset=[0] status=UP)
2022-01-21T21:30:23.749159Z sched-fluxion-resource.debug[0]: resource status changed (rankset=[1-2] status=UP)
2022-01-21T21:30:23.750069Z sched-fluxion-qmanager.debug[0]: handshaking with job-manager completed
2022-01-21T21:30:24.132236Z sched-fluxion-resource.debug[0]: resource status changed (rankset=[3] status=UP)
2022-01-21T21:30:26.490982Z sched-fluxion-resource.debug[0]: status_request_cb: status succeeded
2022-01-21T21:30:26.491095Z sched-fluxion-qmanager.debug[0]: status_request_cb: resource-status succeeded
2022-01-21T21:30:33.294255Z sched-fluxion-qmanager.debug[0]: alloc success (queue=default id=181445591040)
2022-01-21T21:30:33.351415Z sched-fluxion-qmanager.debug[0]: free succeeded (queue=default id=181445591040)
2022-01-21T21:30:54.871116Z sched-fluxion-qmanager.debug[0]: alloc success (queue=default id=543414026240)
2022-01-21T21:30:54.871908Z sched-fluxion-qmanager.debug[0]: alloc success (queue=default id=543414026241)
2022-01-21T21:30:54.922304Z sched-fluxion-qmanager.debug[0]: free succeeded (queue=default id=543414026240)
2022-01-21T21:30:54.926329Z sched-fluxion-qmanager.debug[0]: free succeeded (queue=default id=543414026241)
```

**Comment by dongahn** [comment]
@grondo: I think I repro this on my dev system. 

>> But when I request 2 tasks, the assigned resource set seems to unexpectedly have 2 nodes:

> Hmmm. This should be scheduled to a single node. I will take a look.

This problem only occurs with `sched-fluxion-module`. So this is work yet to be done since I have only been testing this w/ `resource-query`.

Good new is `resource-query` at least behaves correctly! I hope this is straightforward to fix.

Thank you for poking at it. This gives me an area to focus!!

**Comment by dongahn** [comment]
This appears to be an issue with the hwloc reader. It currently doesn't not split the self-discovered hostname into the base name and ID. So when the new match policies use the node ID as the main scoring factor, all nodes are scored the same so the slots are spread. I have a reasonable fix. Thanks again.

**Comment by dongahn** [comment]
@grondo and @milroy: this PR is getting bigger. What is in there is actually a good cutting point. I took off the WIP tag from the title and also added a better description. 

@grondo: do you think you can try this version of fluke to see if this behaves better in this LC system instance environment?

**Comment by grondo** [comment]
> @grondo: do you think you can try this version of fluke to see if this behaves better in this LC system instance environment?

This branch is working well for me in my tests on fluke now.

I haven't yet found any cases where the allocated resource set is not what I'd expect.

Reminder: I'm launching my tests with
```console
$ flux mini alloc \
    --env=FLUX_MODULE_PATH_PREPEND=$(pwd)/qmanager/modules/.libs:$(pwd)/resource/modules/.libs \
    --env=FLUXION_RESOURCE_OPTIONS="policy=hinodex load-allowlist=node,core,gpu" -n 16
```

**Comment by grondo** [comment]
@dongahn: Here's some quick sanity tests I ran against this branch. The main thing I was testing is that some of the boundary conditions are handled (request for ncores <= 4 gets one whole node, 5 cores gets two whole nodes, etc.)

```
ƒ(s=4,d=1) grondo@fluke96:~/git/flux-sched.git$ flux job info $(flux mini submit -n1 hostname) R | jq .execution.R_lite
[
  {
    "rank": "3",
    "children": {
      "core": "0-3"
    }
  }
]
ƒ(s=4,d=1) grondo@fluke96:~/git/flux-sched.git$ flux job info $(flux mini submit -n3 hostname) R | jq .execution.R_lite
[
  {
    "rank": "3",
    "children": {
      "core": "0-3"
    }
  }
]
ƒ(s=4,d=1) grondo@fluke96:~/git/flux-sched.git$ flux job info $(flux mini submit -n4 hostname) R | jq .execution.R_lite
[
  {
    "rank": "3",
    "children": {
      "core": "0-3"
    }
  }
]
ƒ(s=4,d=1) grondo@fluke96:~/git/flux-sched.git$ flux job info $(flux mini submit -n5 hostname) R | jq .execution.R_lite
[
  {
    "rank": "2-3",
    "children": {
      "core": "0-3"
    }
  }
]
ƒ(s=4,d=1) grondo@fluke96:~/git/flux-sched.git$ flux job info $(flux mini submit -n5 -N1 hostname) R | jq .execution.R_lite
flux-job: job 6289509842944 id or key not found
ƒ(s=4,d=1) grondo@fluke96:~/git/flux-sched.git$ flux job attach 6289509842944
0.015s: job.exception type=alloc severity=0 alloc denied due to type="unsatisfiable"
```

```

File: resource/policies/dfu_match_multilevel_id_impl.hpp [hunk 1]
```diff
@@ -34,14 +34,39 @@ multilevel_id_t<FOLD>::score_factor_t::score_factor_t (const std::string &type,
...
+                                    int64_t break_tie) const
 {
-    return (base_factor + m_add_by) * m_multiply_by;
+    int64_t add, mul, tie, factor;
+    if (base_factor < 0 || break_tie < 0) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    add = base_factor + m_add_by;
```
Comment (milroy) [inline resource/policies/dfu_match_multilevel_id_impl.hpp:47]: I think this results in undefined behavior if the overflow occurs.  According to [this paper](https://www.cs.utah.edu/~regehr/papers/overflow12.pdf), it's better to use a precondition check. For example, signed addition will wrap _**if and only if**_ this expression is true (s1 > 0) ∧ (s2 > 0)∧(s1 > (INT_MAX − s2)))∨((s1 < 0) ∧ (s2 < 0)∧(s1 < (INT_MIN − s2))).  See Section III B1 and B3 (for a good alternative).

In the case here, s1 = `base_factor` and s2 = `m_add_by`.

The operation independent check uses extended arithmetic:
```
result = extend(s1) op extend(s2)
if (result < MIN || result > MAX) then
    call failure handler
endif
```

File: t/t3003-resource-global.t [hunk 2]
```diff
@@ -49,4 +49,53 @@ test_expect_success "${test003_desc}" '
     test_cmp 003.R.out ${exp_dir}/003.R.out
```
Comment (milroy) [inline t/t3003-resource-global.t:49]: Nit in commit message: "testsuite: node-exclusive for global **contraints**"

File: t/t3003-resource-global.t [hunk 3]
```diff
@@ -49,4 +49,53 @@ test_expect_success "${test003_desc}" '
...
+#     For a jobspec with node[1]->slot[1]->core[1], it selects
+#     36 cores from the selected node if there is a total of
+#     36 cores in that node.
+#
+#     For a jobspec with slot[18]->core[1], it selects
+#     again all 36 cores from the current available highest node.
+#
+
+cmds004="${cmd_dir}/cmds01.in"
+test004_desc="match allocate with different rack contraints (pol=hinodex)"
```
Comment (milroy) [inline t/t3003-resource-global.t:70]: Nit: typo (contraints)

File: t/data/resource/expected/mt-storage/L2/011.R.out [hunk 4]
```diff
@@ -0,0 +1,264 @@
+      ---------------core0[1:x]
```
Comment (milroy) [inline t/data/resource/expected/mt-storage/L2/011.R.out:1]: Nit typo in commit message: "**Jode**-exclusive match policies"

File: t/t3033-resource-nodex.t [hunk 5]
```diff
@@ -0,0 +1,92 @@
+#!/bin/sh
```
Comment (milroy) [inline t/t3033-resource-nodex.t:1]: Nit in commit message: contraints

**Review by milroy — CHANGES_REQUESTED** [review]
This PR is very impressive.  

Many of my changes are typo nits in commit messages.  Only one of the typo-related changes actually affects comprehension, though.  There are some indentation problems in `resource/traversers/dfu_impl.cpp` that should be fixed.

I believe I found a problem with the overflow detection of `multilevel_id_t<FOLD>::score_factor_t::calc_factor` which should be addressed.

File: resource/traversers/dfu_impl.cpp [hunk 6]
```diff
@@ -798,41 +798,123 @@ int dfu_impl_t::dom_find_dfv (std::shared_ptr<match_writers_t> &w,
...
+    const std::string &dom = m_match->dom_subsystem ();
+    f_out_edg_iterator_t ei, ei_end;
+    (*m_graph)[u].idata.colors[dom] = m_color.gray ();
+
+    for (auto &subsystem : m_match->subsystems ()) {
+        for (tie (ei, ei_end) = out_edges (u, *m_graph); ei != ei_end; ++ei) {
+            if (!in_subsystem (*ei, subsystem) || stop_explore (*ei, subsystem))
+                continue;
+	    vtx_t tgt = target (*ei, *m_graph);
+            bool tgt_constrained = true;
```
[the remark this answers is not in the document]
Comment (dongahn) [reply resource/traversers/dfu_impl.cpp:813]: Thanks! My new native ubuntu development desktop with vim setting is causing these errors... Sorry.

File: resource/policies/dfu_match_multilevel_id_impl.hpp (hunk 1 above)
Comment (dongahn) [reply resource/policies/dfu_match_multilevel_id_impl.hpp:47]: Great. I was looking for a reference for this and could only find blog-quality ones!

**Comment by dongahn** [comment]
@milroy: thank you for your review! I believe I addressed all of your change requests and forced a push. 

If you are okay, please go ahead and approve and put MWP. Thanks!

**Comment by milroy** [comment]
LGTM! Adding MWP.

**Comment by dongahn** [comment]
@milroy: this is huge! Thanks!
